File: Graph/go/best_bridge.go
package graph

type bridgeState struct {
	row, col int
	dist     int
}

// BestBridge returns the minimum number of water tiles to flip to connect two islands.
func BestBridge(grid [][]string) int {
	firstIsland := map[string]struct{}{}
	found := false

	for r := range grid {
		for c := range grid[0] {
			if grid[r][c] == "L" {
				collectIsland(grid, r, c, firstIsland)
				if len(firstIsland) > 0 {
					found = true
					break
				}
			}
		}
		if found {
			break
		}
	}

	if len(firstIsland) == 0 {
		return 0
	}

	visited := map[string]struct{}{}
	queue := []bridgeState{}
	for pos := range firstIsland {
		r, c := parsePosition(pos)
		queue = append(queue, bridgeState{row: r, col: c, dist: 0})
		visited[pos] = struct{}{}
	}

	deltas := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		state := queue[0]
		queue = queue[1:]

		pos := positionKey(state.row, state.col)
		if grid[state.row][state.col] == "L" {
			if _, inside := firstIsland[pos]; !inside {
				return state.dist - 1
			}
		}

		for _, delta := range deltas {
			nr := state.row + delta[0]
			nc := state.col + delta[1]
			if !inBounds(grid, nr, nc) {
				continue
			}
			neighborPos := positionKey(nr, nc)
			if _, seen := visited[neighborPos]; seen {
				continue
			}
			visited[neighborPos] = struct{}{}
			queue = append(queue, bridgeState{row: nr, col: nc, dist: state.dist + 1})
		}
	}

	return -1
}

func collectIsland(grid [][]string, r, c int, island map[string]struct{}) {
	if !inBounds(grid, r, c) {
		return
	}
	if grid[r][c] == "W" {
		return
	}
	pos := positionKey(r, c)
	if _, seen := island[pos]; seen {
		return
	}
	island[pos] = struct{}{}

	collectIsland(grid, r-1, c, island)
	collectIsland(grid, r+1, c, island)
	collectIsland(grid, r, c-1, island)
	collectIsland(grid, r, c+1, island)
}

File: Graph/php/bestBridge.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/GraphUtils.php';

function bestBridge(array $grid): int
{
    $firstIsland = [];

    for ($r = 0; $r < count($grid); $r++) {
        for ($c = 0; $c < count($grid[0]); $c++) {
            if ($grid[$r][$c] === 'L') {
                collectIsland($grid, $r, $c, $firstIsland);
                if ($firstIsland) {
                    break 2;
                }
            }
        }
    }

    if (!$firstIsland) {
        return 0;
    }

    $queue = [];
    $visited = $firstIsland;
    foreach (array_keys($firstIsland) as $key) {
        [$row, $col] = array_map('intval', explode(',', $key));
        $queue[] = [$row, $col, 0];
    }

    $deltas = [[-1, 0], [1, 0], [0, -1], [0, 1]];

    while ($queue) {
        [$row, $col, $dist] = array_shift($queue);
        $key = gridPosKey($row, $col);
        if ($grid[$row][$col] === 'L' && !isset($firstIsland[$key])) {
            return $dist - 1;
        }

        foreach ($deltas as [$dr, $dc]) {
            $nr = $row + $dr;
            $nc = $col + $dc;
            $neighborKey = gridPosKey($nr, $nc);
            if (!gridInBounds($grid, $nr, $nc) || isset($visited[$neighborKey])) {
                continue;
            }
            $visited[$neighborKey] = true;
            $queue[] = [$nr, $nc, $dist + 1];
        }
    }

    return -1;
}

function collectIsland(array $grid, int $r, int $c, array &$island): void
{
    if (!gridInBounds($grid, $r, $c) || $grid[$r][$c] === 'W') {
        return;
    }

    $key = gridPosKey($r, $c);
    if (isset($island[$key])) {
        return;
    }
    $island[$key] = true;

    collectIsland($grid, $r - 1, $c, $island);
    collectIsland($grid, $r + 1, $c, $island);
    collectIsland($grid, $r, $c - 1, $island);
    collectIsland($grid, $r, $c + 1, $island);
}
